### ISSUE TITLE
FlipVR Controllers Do Not Track Position/Rotation

### ISSUE THREAD

**Issue by PointerOffset** [Yellow-Dog-Man/Resonite-Issues#2606]
### Describe the bug?

When using FlipVR controllers in Resonite the users hands will stay at the users's playspace root and will not track position or rotation. Buttons and hand poses appear to work through Quest Touch fallback support.

I am filing this as a bug since the behavior appears distinctly different compared to other controllers that do not have explicit support in Resonite. Most controllers have button mapping issues while the lack of tracking _appears_ to be some kind of other failure.

- Replicated using Playspace Calibrator on a Quest Pro w/ Steam Link. 
- Also replicated with a Valve Index as native PCVR.
- I disabled all overlays and add-ons in SteamVR.

### To Reproduce

- Install the ShiftAll Controller Drivers through Steam: https://store.steampowered.com/app/3036810/Shiftall_Controller_Driver/
- Pair the FlipVR controllers through SteamVR.
- Start up SteamVR
- Start up Resonite

### Expected behavior

Controllers will track position and rotation, operating with Quest Touch fallback controls/skeleton.

### Resonite Version Number

2024.7.17.1173

### What Platforms does this occur on?

Windows

### What headset if any do you use?

Quest Pro (Steam Link), Valve Index

### Log Files

[DESKTOP-3NG8JGV - 2024.7.17.1173 - 2024-07-17 14_16_36.log](https://github.com/user-attachments/files/16275123/DESKTOP-3NG8JGV.-.2024.7.17.1173.-.2024-07-17.14_16_36.log)

### Additional Context

As these controllers are new, I do not expect anyone on the team has them. I am able lend my pair to the development team for troubleshooting if requested.

If required, all relevant resources for the FlipVR controller can be retrieved by installing the ShiftAll Controller Driver via Steam. Includes profiles, render models, etc.

https://store.steampowered.com/app/3036810/Shiftall_Controller_Driver/

Potentially Related Issues:
- https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/2572
- https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/468
- https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/191
- https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/312
- https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/2425
- https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/2568

### Reporters

@PointerOffset / Spex

**Comment by PointerOffset** [comment]
Adding `-LegacySteamVRInput` to my launch arguments fixes the lack of positional tracking. Unfortunately the hand skeleton is still incorrect of course. My hands are rotated 90 degrees (as if they are laid flat on a table) and there is no skeletal animation.

This is much closer to the behavior I would expect from an unsupported controller. It might be appropriate to convert this to a feature request.

**Comment by shiftyscales** [comment]
> 2:16:55 PM.532 (  0 FPS)	Cannot bind FrooxEngine.GlobalActions to VR Controllers

Yes, this issue would be better suited as a feature request / hardware integration. The lack of hand poses / skeletal animation suggests that the skeletal data is not assigned / forwarded from the SteamVR input profile in Resonite.

**Comment by PointerOffset** [comment]
I want to make a note here as I've been investigating what I can do to improve these controllers for myself in Resonite while I wait for more direct support.

The issue with the skeleton input may not be a bug in passing the data. The FlipVR driver just may not be providing skeletal data in the first place. For them, this made sense, as the controller was very much designed with VRChat in mind and Skeletal Input wasn't required. However, VRChat's latest Beta now supports Skeletal input and has caused similar problems with this controller on that platform during testing.

I asked ShiftAll in their Discord and they acknowledged skeletal input is a problem. It's not clear from the translation if they have a bug preventing it from working or if they simply don't support it. Thankfully they at least appear motivated to resolve it.
https://discord.com/channels/1093765691582185482/1266158155855958046/1266398613496401941
<img width="834" alt="image" src="https://github.com/user-attachments/assets/e07b8fca-aab4-486e-bbbf-4a7ae38e1870">

**TL;DR**: These controllers just may not have skeletal input data available right now. It's not clear if that will be resolved by ShiftAll, but hopefully so with VRChat now supporting Skeletal Input too.

**Comment by dragos240** [comment]
I just got these controllers today. It would be great to at least get position and rotation... because my hands are by their sides always and it's really not usable rn.

**Comment by dragos240** [comment]
I just got these controllers today. It would be great to at least get position and rotation... because my hands are by their sides always and it's really not usable rn.

**Comment by dragos240** [comment]
So something that strikes me as odd is when I include `-LegacySteamVRInput`, my lasers work correctly, but my hands do not follow me. They are either at my sides or in the air, but they don't move when my lasers do.

**Comment by PointerOffset** [comment]
I've been trying to get these controllers running well under the Monado OpenXR runtime for Linux. I've actually been successful and during that process I found these are, effectively, "Touch controllers with Index controller offsets."

There is probably a little bit of tweaking that should be done to the offsets but you can basically bind the inputs like a Touch controller and provide positional/rotational offsets like an Index controller. This should make for an easy starting point.

**Comment by TumnusB** [comment]
> I've been trying to get these controllers running well under the Monado OpenXR runtime for Linux. I've actually been successful and during that process I found these are, effectively, "Touch controllers with Index controller offsets."
> 
> There is probably a little bit of tweaking that should be done to the offsets but you can basically bind the inputs like a Touch controller and provide positional/rotational offsets like an Index controller. This should make for an easy starting point.

For Linux this may help
https://lvra.gitlab.io/docs/fossvr/monado/flipvr/

Edit: just noticed it was you who implemented this nvm

**Comment by PointerOffset** [comment]
I wanted to mention here as of January 2nd, my changes to the Monado OpenXR runtime have been merged to their main repository. So FlipVR controllers should now "just work" with Resonite if you are using the latest build of Monado. Keep in mind you'll still need to install ShiftAll's Controller Driver on Steam and symlink the driver folder to the correct location for it to work. The page linked above has up-to-date information: 
https://lvra.gitlab.io/docs/fossvr/monado/flipvr/

So for Resonite users on Linux+Monado, they're no longer blocked on using FlipVR controllers. Windows users are still unable to use them though and I have heard using the `-LegacySteamVRInput` argument has stopped working after the splittenning. I haven't had a chance to test this though.

> It's a workaround and not really a solution, because it just makes Resonite accept the controller inputs as Quest controller inputs

To be clear, this isn't a workaround. Were the controllers functioning as expected in SteamVR, they would still be emulating Quest controllers. This is how FlipVR controllers are expected to be used with any games that don't have explicit support for them. Without official support in Resonite, this is pretty much the "correct" solution.
